### PR TITLE
feat: add nix build infrastructure + github workflow for tests

### PR DESCRIPTION
**Pull request by mrcjkb**
Hi.

I am looking into adding support for haskell-debug-adapter to my [haskell-tools.nvim neovim plugin](https://github.com/mrcjkb/haskell-tools.nvim).
This means I may become interested in contributing to this project in the future. :smile: 

As I use [NixOS](https://nixos.org/) for development, I have added some infrastructure for using nix to build this project and run tests - which I am requesting to merge with this PR.

I have also added a GitHub action that runs the test suite on PR and push to `master`.

The nix flake uses [pre-commit-hooks.nix](https://github.com/cachix/pre-commit-hooks.nix) to

- run `hpack` to generate a cabal file from the `package.yaml`.
- run [`cabal2nix`](https://github.com/NixOS/cabal2nix) to generate a `default.nix` from the previously generated cabal file - which is then used in `nix/overlay.nix` to build the package.
- run [`alejandra`](https://github.com/kamadorueda/alejandra) for nix formatting.

This could potentially be extended to also run `hlint` and a Haskell formatter. I have omitted these, as they would result in modifications to the Haskell source files.
A potential downside to adding linters and formatters is that the CI could fail if someone uses a different version of `hlint` or the formatter than the version used in the nix build.
For anyone who has nix installed, this is not a problem, as long as they use `nix develop` to enter a nix shell with the correct versions of the packages installed.

The same applies to the `haskell-debug-adapter.cabal`, which is generated by `hpack`. The file has been changed in this PR, because the nix infrastructure uses a newer version of `hpack`.

`hpack` has lowered the lower bounds of the dependencies in `haskell-debug-adapter.cabal`.
This indicates that the `package.yaml` and stack resolver is not the source of truth for this build (?).
If need be, I can revert that change and remove the `hpack` pre-commit-hook. Although this discrepancy looks to me like it could lead to inconsistent results between stack and cabal builds.

### CODE REVIEW

**Comment by mrcjkb** [comment]
Marking this as a draft.
It might be better to use the locally checked out repositories as inputs, instead of the dependencies on hackage.

**Comment by phoityne** [comment]
Thank you for your contribution. 
I'm going to check building with your cabal file locally.

**Comment by mrcjkb** [comment]
@phoityne I have added the nix infrastructure to https://github.com/phoityne/haskell-dap/pull/14 and https://github.com/phoityne/ghci-dap/pull/11.

They are used as inputs for this flake, allowing for hermetic builds during development.
If either of the dependencies is changed, the changes can be applied to this project with `nix flake update`.

Note: IMO, these projects would be easier to maintain if they were merged into a monorepo.
Is there a reason you have chosen to put them in separate repositories?

File: flake.nix [hunk 1]
```diff
@@ -0,0 +1,107 @@
+{
+  description = "Debug Adapter for Haskell debugging system.";
+
+  inputs = {
+    nixpkgs.url = "github:nixos/nixpkgs/nixpkgs-unstable";
+
+    # TODO: Replace with upstream url when merged
```
Comment (mrcjkb) [inline flake.nix]: - [ ] TODO

**Comment by phoityne** [comment]
Hi.
I'm not a NixOS user.
Is this PR what haskell-tools.nvim needs ?
PR to build haskell-tools.nvim on haskell-tools.nvim repository ?

Or an improvement suggestion for this project ?
(such as auto lint, build, test,, by github actions.)

**Comment by mrcjkb** [comment]
It's not needed for haskell-tools.nvim.

* It would significantly improve my ability to contribute to this repository if I need to.
* It also makes it easy for nix or NixOS users to install haskell-debug-adapter from source.

(nix can be used on any Linux distro and on MacOS).

The GitHub actions workflow is an improvement suggestion.

**Comment by phoityne** [comment]
Thank you for the suggestion.
Is this still draft state ?
Because no side effect to users,  released packages, haskell-tools.nvim, 
I think it's ok to be merged.

**Comment by mrcjkb** [comment]
> Thank you for the suggestion. Is this still draft state ? Because no side effect to users, released packages, haskell-tools.nvim, I think it's ok to be merged.

I just have to update the flake inputs for ghci-dap and haskell-dap to point to your repos (they currently point to my forks) and update the lock file. Then this should be ready.

I'll do that later today.

**Comment by mrcjkb** [comment]
Done. This is ready to be merged :)

**Comment by phoityne** [comment]
Thanks for your contribution.
merged.

In the future, I hope to be able to use the same mechanism to check builds and create binaries.

Especially ghci-dap,
I think you can install it with ghcup if we prepare binaries for each distro and ghc version.

**Comment by mrcjkb** [comment]
Sounds reasonable.

If ghci-dap is a requirement for haskell-debug-adapter, it might make sense to package them together.

The nice thing about using nix for tests/checks is that they can be reproduced locally by anyone who has nix installed.
That can often be difficult with GitHub actions. But for this project, it definitely makes sense to test with different ghc versions, too.

**Pull request merged**